requirements-lint.in

```diff
@@ -0,0 +1,2 @@
+prospector==1.1.6.2
+-r requirements-tests.txt
```

Comment:
I think it makes sense for `make lint` to have its own separate requirements file, even though it has to install all the test requirements in its venv (because pylint uses them). It means that `make lint` and `make tests` are completely independent of each other and can run in parallel.

Reply:
What this means on CI is that `make tests` and `make lint` can run simultaneously on separate Jenkins nodes. They'll be duplicating some work by both installing `requirements-tests.txt`. But after that they'll each be doing separate work. The simplicity of having the two be totally independent of each other enables full concurrency to be used in a simple way.

The alternative would be to create only one venv, install `requirements-test.txt` into it once, and then run the linter and then the tests in serial. Or, I think after installing `requirements-test.txt` we could run linter and tests in parallel but on the same Jenkins node. I think this would use less CPU cycles but take longer to run to completion because it's not able to use parallelism as well.

There might be something that we can do with Docker where we create one venv, install `requirements-text.txt` into it once, then take a Docker image of that, then run `make tests` and `make lint` in parallel, even on separate Jenkins nodes, both starting from that Docker image we just took. But this seems too complicated to me.

I think I lean towards spending more CPU cycles (which with Jenkins slave nodes, or just a big enough single Jenkins machine, shouldn't cause resource competition between jobs) in order to get the most parallelism in the simplest way and have the Jenkins pipeline finish as soon as possible.

Reply:
In my experiments in https://github.com/hypothesis/lms/pull/686 I struck a balance by having two separate backend tasks: one for running tests + coverage and one for everything else (lint, formatting etc.). In that PR, lint, formatting etc. all still use separate tox envs. I suspect that if you combined the lint/formatting/docstrings tox envs, that build would complete much faster. In fact, I might give that a try.

For the frontend tasks everything was fast enough that it made sense to run it was one task. As you can see from the screenshots on that PR, the overall build actually completed faster this way given our current resources.

Reply:
Backend tests + coverage I agree with: we should merge the `coverage`, `codecov` and `clean` tox factors into the `tests` factor, so there will only be one virtualenv and one requirements file (the "tests" venv and requirements file) used for all of tests, coverage, codecov and clean together. And then Jenkins will continue to run `make backend-tests coverage codecov` as a single, serial step as it already does on master. (These three have to be run in serial anyway since `codecov` cannot work if `coverage` hasn't been run to completion before it, and `coverage` in turn depends on `backend-tests`.)

Note that this change won't actually make much difference to Jenkins however, because  `make backend-tests coverage codecov` is already run in serial on Jenkins and, while the change would cause the three to share a single virtualenv instead of making three virtualenvs, there aren't many duplicate dependencies being installed into these three separate virtualenvs currently anyway: the production and tests deps (including `coverage`) get installed into the `tests` venv; only `coverage` gets installed into the `coverage` venv (so the `coverage` package is being installed twice in two different venvs, but that's just one package and its dependencies); and only `codecov` gets installed into the `codecov` venv. So we _would_ remove some package install duplication here; and creating one venv instead of three will save venv-creation time; but the gain is going to be minimal here.

A big part of the reason for running things in separate venvs is that it enables them to be parallelised in a simple way. But `backend-tests`, `coverage` and `codecov` are _not_ parallelizable as they depend on each other's outputs, so the benefit is null. The other big reason for separate venvs is that separating deps reduces the chances of dependency collisions. But `coverage` already has to be installed in the `tests` venv anyway (since it has to be used to wrap the `pytest` command) so the only dependency being isolated is `codecov`, so that benefit is nearly mute as well.

So these make little sense as separate venvs basically!

Reply:
Combining `lint`, `checkformatting` and others together I'm not sure about. `checkformatting` shares no requirements with `lint`, so there's no duplicate package installs to be avoided here, and `checkformatting` is a really small and fast one (it only has to install and run `black`, which is much faster than what `lint` or tests` have to do), and I think it's nice to let it run unencumbered on its own because if `black` is going to fail it will fail fast (before `tests` or `lint` have even finished installing their requirements, probably) and then when it fails any other parallel stages still running will be aborted.

There might be more duplication to be removed by combining `checkformatting` and `lint` together but I think changing to Jenkins slave nodes (which David says should be easy to do and is scheduled for next sprint) might change things so that running them in parallel rather than serial is better,

Reply:
Another obvious one to consider is combining `tests` and `lint`. There _is_ a lot of duplicate package installing to be saved by combining those two into one. They both require all of the production and tests dependencies (as does `checkdocstring` IIRC). But they also both take kind of a long time to run after the deps have been installed, so I'm unclear whether saving CPU cycles by running them in serial is going to be better than spending a lot more CPU cycles but running them in parallel which I think should finish a lot faster (at least if they don't all slow each other down by competing for resources).